refactor: Refactor part 2 to remove additional remaining api calls to LMS via REST

binodpant opened the pull request:
## Includes

Update:
Going to do a separate PR for this one:
   - [ ] Call lms util instead of REST api to get assessment grades (current call is `grades_api.get_course_assessment_grades`)
 The above needs a edx-platform PR to be merged first

This PR only includes the following:
- [x] Remaining refactors from https://github.com/edx/edx-enterprise/pull/1199
   - [x] Call lms util instead of REST Api for `get_course_details` , this one uses the `CourseApiClient` whose usage can also be removed from this codebase
- [x] Bug fix for bug found in staging testing and test

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

binodpant commented on integrated_channels/integrated_channel/exporters/learner_data.py:
```diff
@@ -87,22 +87,6 @@ def grade_audit(self):
         """
         return self.GRADE_AUDIT
 
-    def fetch_course_details(self, course_details, enrollment_course_id):
```
this caching was not being used so simplified it by just calling the new method now

binodpant commented on integrated_channels/lms_utils.py:
```diff
@@ -51,5 +62,25 @@ def get_single_user_grade(course_id, user):
     return course_grade
+
+
+def get_course_details(course_id):
```
I need to add a test for this method, will get that in soon

edx-abolger replied:
Looks like you have added one, resolving this comment.

binodpant commented on tests/test_integrated_channels/test_lms_utils.py:
```diff
@@ -51,6 +52,10 @@ def test_get_single_user_grade_success(self, mock_course_grade_factory):
         mock_course_grade_factory.return_value.read.return_value = expected_grade
         single_user_grade = get_single_user_grade(A_GOOD_COURSE_ID, self.user)
         assert single_user_grade == expected_grade
+        mock_course_grade_factory.return_value.read.assert_called_with(
```
this test covers the bug fix for a bug I found today in prior PR

edx-abolger reviewed (commented):
Will wait on additional tests before approving.

binodpant commented on integrated_channels/lms_utils.py:
```diff
@@ -14,6 +15,16 @@
+
 
 def get_course_certificate(course_id, user):
     """
```
Ah I double checked, so yes it is used like follows, either it throws that key error or returns an invalid cert

```
try:
            certificate = get_course_certificate(course_id, user)
        except InvalidKeyError:
            self._log_courseid_not_found(course_id, enterprise_enrollment, lms_user_id)
            return completed_date, grade, is_passing, percent_grade

        if not certificate:
            self._log_cert_not_found(course_id, enterprise_enrollment, lms_user_id)
            return completed_date, grade, is_passing, percent_grade
```


the impl in edx-platform of `get_certificate_for_user` looks like this so yeah looks like we are covered? 

```
try:
        cert = GeneratedCertificate.eligible_certificates.get(
            user__username=username,
            course_id=course_key
        )
    except GeneratedCertificate.DoesNotExist:
        return None

    if format_results:
        return format_certificate_for_user(username, cert)
    else:
        return cert
```

edx-abolger replied:
I meant the docstring says one thing and the code does something else. Also the code behaves differently under 2 different error conditions, which means using this function will require the caller to handle None and also handle an exception. Can you make the function more consistent?

binodpant replied:
yeah that case is interesting since it's due an importError. Perhaps it's best indicated if I actually threw an ImportError

```
if not get_certificate_for_user:
        raise ImportError("Could not load get_certificate_for_user function")
```

binodpant replied:
Ah I see we have this pattern in other plcaes 

```
if get_certificate_for_user is None:
            raise NotConnectedToOpenEdX(
                _('To use this EnterpriseCourseEnrollmentSerializer, this package must be '
                  'installed in an Open edX environment.')
            )
```

binodpant replied:
yeah so going with this pattern here too basically the user of this api only has to handle InvalidKeyError now, and updated docstring

edx-abolger replied:
Thank you. This is much more consistent as Python has nice syntax for things like this: 
```

try:
    # do something that may fail
except IDontLikeYouException:
    # say please
except YouAreTooShortException:
    # stand on a ladder
```

binodpant commented on integrated_channels/lms_utils.py:
```diff
@@ -51,5 +62,25 @@ def get_single_user_grade(course_id, user):
     if not CourseGradeFactory:
```
did you mean line 38 with this? 

```
user_cert = get_certificate_for_user(username=user.username, course_key=course_key)
```

checked that method again in edx-platform and indeed it expects username instead of user (odd)

binodpant replied:
```
def get_certificate_for_user(username, course_key, format_results=True):
    """
    Retrieve certificate information for a particular user for a specific course.

    Arguments:
        username (unicode): The identifier of the user.
        course_key (CourseKey): A Course Key.
```

edx-abolger replied:
Ew. not great but obviously outside the scope of this ticket and our team. Will close comment.

binodpant commented:
This PR (the grades api client part) is blocked until I release edx-platform to offer a python api to fetch single learner grade

edx-abolger commented on integrated_channels/lms_utils.py:
```diff
@@ -42,14 +56,44 @@ def get_course_certificate(course_id, user):
+
+def get_course_details(course_id):
+    """
+    Returns:
```
Again, this is not a clean interface. You just moved the problems: you are mixing "Return an error code OR throw an exception, the same 20 line function."  That is 2 different problems the caller now has to handle. Either ALWAYS return none+error code or ALWAYS let the exceptions pass through.

edx-abolger replied:
To explain: the problem with this is that now the calling code will have to write _extra_ code and do _more_ weird error handling, not to mention the code author will have to go look inside this function to figure out how to write the error handler for it.

binodpant replied:
Ah I see what you are saying, so the current usage looks like this with a mixed bag of None vs exceptions

```
course_details = get_course_details(course_run_id)
if course_details is None:
    # Course not found, so we have nothing to report.
```

Instead we can go:

```
try:
    course_details = get_course_details(course_run_id)
except InvalidKeyError:                             # Course not found, log error
except CourseOverview.DoesNotExist:   # Course key is invalid, log error
```

This will also make this function consistent with the other ones in the pr

binodpant replied:
note, the caller has to still handle these two exceptions, since we call this function in a loop of enrollments, we want to log and continue the loop in these two cases. The NotConnectedException is different, it should be treated as runtime exception since something is really bad, no point in continuing. But rather than swallow these exceptions to is better regardless, to propagate them from the get_course_details function.

binodpant replied:
@edx-abolger sorry for the multiple rounds here, but this new pattern look better and more consistent hopefully

binodpant commented:
also I just discovered a mistake I had made. the methods need to process course_key not course_id. The names are confusing me, as used in the code. The enterprise_enrollment.course_id are actually course_keys !

binodpant commented:
and the `course_id` expected by the lms apis are actually instances of CourseKey objects

```
course_id = CourseKey.from_string(course_key)
```

binodpant commented on integrated_channels/lms_utils.py:
```diff
@@ -67,31 +67,29 @@ def get_single_user_grade(course_id, user):
-    course_key = CourseKey.from_string(course_id)
-    course_grade = CourseGradeFactory().read(user, course_key=course_key)
+    course_id = CourseKey.from_string(course_key)
+    course_grade = CourseGradeFactory().read(user, course_key=course_id)
```
note: every thing that says `course_id` in the grading apis refers to instances of CourseKey which have to be obtained via CourseKey.from_string() from the course_key strings provided from learner_data.py (those enterprise_enrollment.course_id are actually course key strings)

binodpant commented:
updated with a test fix, and an actual bug I found when running the task since I had misunderstood the `course_id` params that go into the LMs apis

added a proposal to our proposal listing: we don't have integration level tests right now, I think we will want at least one for each type of transmit task for better lockdown of bugs

alex-sheehan-edx reviewed (approved):
G2G 👍
